### ISSUE TITLE
PYI036 (bad-exit-annotation) triggers on seemingly appropriate annotations

### ISSUE THREAD

**Issue by Sachaa-Thanasius** [astral-sh/ruff#9794]
**Code snippet**
For additional context: Importing from `types` outside of the `TYPE_CHECKING` block doesn't trigger this. However, I'd prefer not actually importing that at runtime, since it can be expensive. Curious if this kind of case is in scope for this rule.

```py
from typing import TYPE_CHECKING

from typing_extensions import Self


if TYPE_CHECKING:
    from types import TracebackType
else:

    class TracebackType:
        pass


class Example:
    def __enter__(self) -> Self:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        return None
```

**Invoked Command**
```bash
ruff core\utils\_test.py --isolated --select PYI036
```

**Relevant pyproject.toml Sections**
Use of the `--isolated` and `--select` flags above makes this redundant, I think?

**Version**
ruff 0.2.0

**Comment by T-256** [comment]
In your case I think you could also do:

```py
from __future__ import annotations
from typing import TYPE_CHECKING

from typing_extensions import Self

if TYPE_CHECKING:
    from types import TracebackType


class Example:
    def __enter__(self) -> Self:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        traceback: TracebackType | None,
) -> None:
        return None
```

**Comment by Sachaa-Thanasius** [comment]
While that's true, I would rather not break type introspection for tools like `typing.get_type_hints()`. Having something in the type checking block's else clause prevents that, even if the type isn't exactly right. Hoping there's a "best of both worlds" option here, where I don't break that and also have ruff's rule recognize what's happening here. If not, though, I understand.

**Comment by CarrotManMatt** [comment]
I am running into this bug when importing the types inside a type-checking block & using quoted annotations rather than `from __future__ import annotations`. A resolution for this bug would be appreciated
